velox/common/memory/tests/SharedArbitratorTest.cpp

```diff
@@ -2747,6 +2747,118 @@ DEBUG_ONLY_TEST_F(SharedArbitrationTest, arbitrateMemoryFromOtherOperator) {
+  std::shared_ptr<core::QueryCtx> joinQueryCtx = newQueryCtx(kMemoryCapacity);
+
+  std::atomic<bool> joinWaitFlag{true};
+  folly::EventCount joinWait;
+  std::atomic<bool> taskPauseWaitFlag{true};
+  folly::EventCount taskPauseWait;
+
+  const auto fakeAllocationSize = kMemoryCapacity - (16L << 20);
+
+  fakeOperatorFactory_->setCanReclaim(false);
+  std::atomic<bool> injectAllocationOnce{true};
+  fakeOperatorFactory_->setAllocationCallback([&](Operator* op) {
+    if (!injectAllocationOnce.exchange(false)) {
+      return TestAllocation{};
+    }
+    auto buffer = op->pool()->allocate(fakeAllocationSize);
+    joinWaitFlag = false;
+    joinWait.notifyAll();
+    // Wait for pause to be triggered.
+    taskPauseWait.await([&]() { return !taskPauseWaitFlag.load(); });
+    return TestAllocation{op->pool(), buffer, fakeAllocationSize};
+  });
+
+  const int numDrivers = 4;
+  std::atomic<int> injectedBuildWait{0};
+  SCOPED_TESTVALUE_SET(
+      "facebook::velox::exec::Driver::runInternal::addInput",
+      std::function<void(Operator*)>(([&](Operator* op) {
+        if (op->operatorType() != "HashBuild") {
+          return;
+        }
+        joinWait.await([&]() { return !joinWaitFlag.load(); });
+      })));
+
+  std::atomic<int> numAppends{0};
+  const std::string errMsg("injected spillError");
+  SCOPED_TESTVALUE_SET(
+      "facebook::velox::exec::SpillState::appendToPartition",
+      std::function<void(exec::SpillState*)>([&](exec::SpillState* state) {
+        if (++numAppends != numDrivers) {
+          return;
+        }
+        VELOX_FAIL(errMsg);
+      }));
+
+  SCOPED_TESTVALUE_SET(
+      "facebook::velox::exec::Task::requestPauseLocked",
+      std::function<void(Task*)>(([&](Task* /*unused*/) {
+        LOG(ERROR) << "task pause wait";
+        taskPauseWaitFlag = false;
+        taskPauseWait.notifyAll();
+      })));
+
+  const auto spillDirectory = exec::test::TempDirectoryPath::create();
+  std::thread joinThread([&]() {
+    auto planNodeIdGenerator = std::make_shared<core::PlanNodeIdGenerator>();
+    VELOX_ASSERT_THROW(
+        AssertQueryBuilder(duckDbQueryRunner_)
+            .queryCtx(joinQueryCtx)
+            .spillDirectory(spillDirectory->path)
+            .config(core::QueryConfig::kSpillEnabled, "true")
+            .config(core::QueryConfig::kJoinSpillEnabled, "true")
+            .plan(PlanBuilder(planNodeIdGenerator)
+                      .values(vectors)
+                      .project({"c0 AS t0", "c1 AS t1", "c2 AS t2"})
+                      .hashJoin(
+                          {"t0"},
+                          {"u0"},
+                          PlanBuilder(planNodeIdGenerator)
+                              .values(vectors)
+                              .project({"c0 AS u0", "c1 AS u1", "c2 AS u2"})
+                              .planNode(),
+                          "",
+                          {"t1"},
+                          core::JoinType::kAnti)
```

Comment:
Is it important to use anti join here? Would inner or other type of join not allow to reproduce the necessary conditions?

Reply:
Not important. Any query that generate a large amount of memory usage is fine.

Reply:
I see. How does this test related to making HashBuild spilling run multi-threaded then? Just trying to understand. Would appreciate a clarification.

Reply:
The change is covered by existing test. This additional test is to make sure that the parallel join path can catch spilling error properly which is new.